sw/device/silicon_creator/lib/cert/asn1.c

```diff
@@ -0,0 +1,181 @@
+  if (size > PTRDIFF_MAX)
+    return kErrorAsn1InvalidArgument;
+  if (state->offset + size > state->size)
+    return kErrorAsn1BufferExhausted;
+  memcpy(state->buffer + state->offset, bytes, size);
+  state->offset += size;
+  return kErrorOk;
+}
+
+rom_error_t asn1_start_tag(asn1_state_t *state, asn1_tag_t *new_tag,
+                           uint8_t id) {
+  new_tag->state = state;
+  RETURN_IF_ERROR(asn1_push_byte(state, id));
+  new_tag->len_offset = state->offset;
+  // We do not yet known how many bytes we need to encode the length. For now
+  // reserve three bytes so we can use the 16-bit encoding. This is then fixed
+  // in asn1_finish_tag.
+
+  // Three bytes: one to hold the number of octets to use
+  // and two to store the length.
+  RETURN_IF_ERROR(asn1_push_bytes(state, (uint8_t[]){0x82, 0, 0}, 3));
+  new_tag->len_size = 3;
+  return kErrorOk;
+}
+
+rom_error_t asn1_finish_tag(asn1_tag_t *tag) {
+  // Robustness check against accidently reusing a finished tag.
+  if (tag->state == NULL)
+    return kErrorAsn1Internal;
+  // Sanity check: asn1_start_tag should have output three bytes.
+  if (tag->len_size != 3)
+    return kErrorAsn1Internal;
+  // Compute actually used length.
+  size_t length = tag->state->offset - tag->len_offset - tag->len_size;
+  // Fixup the length bytes and compute the size of the minimal encoding.
+  size_t final_len_size;
+  if (length <= 0x7f) {
+    // We only need one byte to hold the length.
+    tag->state->buffer[tag->len_offset] = (uint8_t)length;
+    final_len_size = 1;
+  } else if (length <= 0xff) {
+    // We need two bytes to hold the length.
+    tag->state->buffer[tag->len_offset + 0] = 0x81;
+    tag->state->buffer[tag->len_offset + 1] = (uint8_t)length;
+    final_len_size = 2;
+  } else if (length <= 0xffff) {
+    // We need three bytes to hold the length.
+    tag->state->buffer[tag->len_offset + 0] = 0x82;
+    tag->state->buffer[tag->len_offset + 1] = (uint8_t)(length >> 8);
+    tag->state->buffer[tag->len_offset + 2] = (uint8_t)(length & 0xff);
+    final_len_size = 3;
+  } else {
+    // Length too large.
+    return kErrorAsn1Internal;
+  }
+  // If the final length uses less bytes than we initially allocated, we
+  // need to shift all the tag data forward.
+  if (final_len_size != tag->len_size) {
+    for (size_t i = 0; i < length; i++) {
+      tag->state->buffer[tag->len_offset + final_len_size + i] =
+          tag->state->buffer[tag->len_offset + tag->len_size + i];
+    }
+  }
+  // Fix up state offset.
+  tag->state->offset -= tag->len_size - final_len_size;
+  // Hardening: clear out the tag structure to prevent accidental reuse.
+  tag->state = NULL;
+  tag->len_offset = 0;
+  tag->len_size = 0;
+  return kErrorOk;
+}
+
+rom_error_t asn1_push_bool(asn1_state_t *state, bool value) {
+  asn1_tag_t tag;
+  RETURN_IF_ERROR(asn1_start_tag(state, &tag, kAsn1TagNumberBoolean));
+  RETURN_IF_ERROR(asn1_push_byte(state, value ? 0xff : 0));
```

Comment:
This way of using the API is not very intuitive, after calling `asn1_start_tag` and getting a new “object” instance, the most logical for me would be calling a function to add elements to this new object (tag). In that case the `finish_tag` function would not be need if you decide that's ok to evaluate the tag size every time a new element is added.

Reply:
It makes sense, so then that would require to duplicate `asn1_tag_start` to push a tag within a tag, the API would be:
- `asn1_state_start/finish`
- `asn1_state_push_byte`
- `asn1_state_push_tag`
- `asn1_tag_push_tag`
- and then all others are `asn1_tag_push_{boolean,integers,...}`

Reply:
The API that you described is exactly what I have in mind.
Yeah, I agree that it would be harder to autogenerate code. But the certificate will easier to read, I believe. 
Maybe you keep the current API leaving together with the new, or let the new to be done in the future.
All let you decide base on the project timeline.